opm/simulators/wells/StandardWell_impl.hpp

```diff
@@ -2329,6 +2333,31 @@ namespace Opm
             }
         }
 
+        // for producers where all perforations have zero rate we
+        // approximate the perforation mixture using the mobility ratio
+        // and weight the perforations using the well transmissibility.
+        bool all_zero = std::all_of(perfRates.begin(), perfRates.end(), [](double val) { return val == 0.0; });
+        if ( all_zero && this->isProducer() ) {
+            double total_tw = 0;
+            for (int perf = 0; perf < nperf; ++perf) {
+                total_tw += well_index_[perf];
+            }
+            for (int perf = 0; perf < nperf; ++perf) {
+                const int cell_idx = well_cells_[perf];
+                // TODO make a derivative free version of getMobility
+                std::vector<EvalWell> mob(num_components_, {numWellEq_ + numEq, 0.});
+                getMobility(ebosSimulator, perf, mob, deferred_logger);
+                const double well_tw_fraction = well_index_[perf] / total_tw;
+                double total_mobility = 0.0;
+                for (int component = 0; component < num_components_; ++component) {
+                    total_mobility += mob[component].value();
+                }
+                for (int component = 0; component < num_components_; ++component) {
+                    perfRates[perf * num_components_ + component] += well_tw_fraction * mob[component].value() / total_mobility;
+                }
+            }
+        }
+
```

Comment:
With the above modification, will `tot_surf_rate==0` still happens in function `computeConnectionDensities`? I think with this, perfRates will not be all zero anymore? Do I miss something here?

Reply:
Yes it can be zero. If for instance the bottom perforation have zero rates, but the top has a non-trivial rate. This happens often for wells where cross flow is not allowed. For instance in model 2.